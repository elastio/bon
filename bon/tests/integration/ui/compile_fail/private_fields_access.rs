#![deny(warnings)]

mod unsafe_fields_are_private {
    struct Sut;

    #[bon::bon]
    impl Sut {
        #[builder]
        fn sut(self, #[builder(start_fn)] _x1: u32, _x2: u32) {}
    }

    fn _test() {
        let sut = Sut.sut(99);

        // Previously, there was an attempt to generate names for private fields
        // with randomness to ensure users don't try to access them. This however,
        // conflicts with caching in some build systems. See the following issue
        // for details: https://github.com/elastio/bon/issues/218
        let SutSutBuilder {
            __unsafe_private_phantom: _,
            __unsafe_private_named: _,
            // These are public
            _x1: _,
            self_receiver: _,
        } = sut;
    }
}

mod self_receiver_and_start_fn_args_must_be_private {
    pub struct MustBePrivate;

    #[bon::bon]
    impl MustBePrivate {
        #[builder]
        pub fn method(self, #[builder(start_fn)] _x1: u32) {}
    }
}

fn main() {
    let builder = self_receiver_and_start_fn_args_must_be_private::MustBePrivate.method(99);
    let _ = builder.self_receiver;
    let _ = builder._x1;
}
