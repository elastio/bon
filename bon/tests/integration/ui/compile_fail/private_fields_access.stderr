error: use of deprecated field `unsafe_fields_are_private::SutSutBuilder::__unsafe_private_phantom`: this field should not be used directly; it's an implementation detail, and if you access it directly, you may break some internal unsafe invariants; if you found yourself needing it, then you are probably doing something wrong; feel free to open an issue/discussion in our GitHub repository (https://github.com/elastio/bon) or ask for help in our Discord server (https://bon-rs.com/discord)
  --> tests/integration/ui/compile_fail/private_fields_access.rs:20:13
   |
20 |             __unsafe_private_phantom: _,
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> tests/integration/ui/compile_fail/private_fields_access.rs:1:9
   |
1  | #![deny(warnings)]
   |         ^^^^^^^^
   = note: `#[deny(deprecated)]` implied by `#[deny(warnings)]`

error: use of deprecated field `unsafe_fields_are_private::SutSutBuilder::__unsafe_private_named`: this field should not be used directly; it's an implementation detail, and if you access it directly, you may break some internal unsafe invariants; if you found yourself needing it, then you are probably doing something wrong; feel free to open an issue/discussion in our GitHub repository (https://github.com/elastio/bon) or ask for help in our Discord server (https://bon-rs.com/discord)
  --> tests/integration/ui/compile_fail/private_fields_access.rs:21:13
   |
21 |             __unsafe_private_named: _,
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0616]: field `self_receiver` of struct `MustBePrivateMethodBuilder` is private
  --> tests/integration/ui/compile_fail/private_fields_access.rs:41:21
   |
41 |     let _ = builder.self_receiver;
   |                     ^^^^^^^^^^^^^ private field

error[E0616]: field `_x1` of struct `MustBePrivateMethodBuilder` is private
  --> tests/integration/ui/compile_fail/private_fields_access.rs:42:21
   |
42 |     let _ = builder._x1;
   |                     ^^^ private field
